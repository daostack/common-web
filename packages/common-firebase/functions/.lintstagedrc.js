module.exports = {
  '**/*.js?(x)': (filenames) => `eslint ${filenames.join(' ')}`
}
