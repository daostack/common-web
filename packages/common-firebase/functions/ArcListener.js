const { Arc } = require('@daostack/arc.js');
const admin = require('firebase-admin');

const graphHttpLink = 'https://api.thegraph.com/subgraphs/name/daostack/v8_1_exp_xdai'
const graphwsLink = 'wss://api.thegraph.com/subgraphs/name/daostack/v8_1_exp_xdai'

const arc = new Arc({
  graphqlHttpProvider: graphHttpLink,
  graphqlWsProvider: graphwsLink,
});

const db = admin.firestore();

// get all DAOs data from graphql and read it into the subgraph
async function updateDaos() {
  const response = [];
  const daos = await arc.daos().first();
  console.log(`Found ${daos.length} DAOs`);
  for (const dao of daos) {
    const joinAndQuitPlugins = await dao.plugins({where: {name: 'JoinAndQuit'}}).first();
    if (joinAndQuitPlugins.length === 0) {
      // not a properly configured common DAO, skipping
      const msg = `Skipping ${dao.id} as it is not properly configured`;
      console.log(msg);
      response.push(msg)
    } else {
      console.log(`updating ${dao.id}`);
      const joinAndQuitPlugin = joinAndQuitPlugins[0];
      const {
        fundingGoal,
        minFeeToJoin,
        memberReputation
      } = joinAndQuitPlugin.coreState.pluginParams;
      try {
      const daoState = dao.coreState
      const doc = {
        id: dao.id,
        address: daoState.address,
        memberCount: daoState.memberCount,
        name: daoState.name,
        numberOfBoostedProposals: daoState.numberOfBoostedProposals,
        numberOfPreBoostedProposals: daoState.numberOfPreBoostedProposals,
        numberOfQueuedProposals: daoState.numberOfQueuedProposals,
        register: daoState.register,
        // reputationId: reputation.id,
        reputationTotalSupply: parseInt(daoState.reputationTotalSupply),
        fundingGoal: fundingGoal.toString(),
        minFeeToJoin: minFeeToJoin.toString(),
        memberReputation: memberReputation.toString()
      }
      await db.collection('daos').doc(dao.id).set(doc)
      const msg =`Updated dao ${dao.id}`
      response.push(msg)
      console.log(msg)
    } catch(err) {
      console.log(err)
      throw err
    }
    }
  }
  return response.join('\n')
}

async function updatePlugins() {

}

async function updateProposals() {
  arc.proposals({}, {subscribe: true, fetchAllData: true})
    .subscribe(async proposals => {
      proposals.map(async proposal => {
        const {coreState} = proposal;
        const proposalObject = {

          boostedAt: coreState.boostedAt,
          createdAt: coreState.createdAt,
          dao: coreState.dao.id,
          description: coreState.description,
          expiresInQueueAt: coreState.expiresInQueueAt,
          executionState: coreState.executionState,
          executed: coreState.executed,
          funding: coreState.funding.toString(),
          executedAt: coreState.executedAt,
          joinAndQuit: {
            funding: coreState.funding.toString(),
            proposedMemberId: coreState.proposer,
            proposedMemberAddress: coreState.proposer,
          },
          preBoostedAt: coreState.preBoostedAt,
          id: coreState.id,
          name: coreState.name,
          proposer: coreState.proposer,
          proposerId: coreState.proposer,
          resolvedAt: coreState.resolvedAt,
          stage: coreState.stage,
          title: coreState.title,
          type: coreState.type,
          votes: {
            list: coreState.votes,
            votesAgainst: parseInt(coreState.votesAgainst),
            votesFor: parseInt(coreState.votesFor),
          },
          links: [{
            title: "website",
            url: coreState.url,
          }],
          images: [],
          winningOutcome: coreState.winningOutcome,
        }
        await db.collection('proposals').doc(coreState.id).set(proposalObject)
      });

    })
}

// function updateDaosSubscription() {
//   //loop that runs a function every 15 seconds for 3 intervals
//   const response = ['xxx']
//   try {
//     const db = admin.firestore();
//     arc
//       .daos({}, {subscribe: true, fetchAllData: true})
//       .subscribe(res => {
//         res.map(async (dao, i) => {
//           // const state = await dao.fetchState()
//           // console.log(state)
//           const {id, address ,
//             memberCount,
//             name,
//             numberOfBoostedProposals ,
//             numberOfPreBoostedProposals ,
//             numberOfQueuedProposals ,
//             register,
//             reputation,
//             reputationTotalSupply,
//           } = dao.coreState;
//           // if (!dao.coreState.name.includes('Test DAO') && !dao.coreState.name.includes('Car DAO')) {
//           if (true) {
//             const joinAndQuitPlugins = await dao.plugins({where: {name: 'JoinAndQuit'}}).first()
//             if (joinAndQuitPlugins.length == 0) {
//               // we'll delete this DAO, as it does not have the correct plugins
//               try {
//                 console.log('DELETING: ', id)
//                 response.append(`Deleted dao ${id}`)
//                 await db.collection('daos').doc(id).delete();
//               } catch (e) {
//                 console.error('Failed to updated DAOs: ', error);
//               }
//             } else {
//               console.log('JaQ PLUGINS: ', joinAndQuitPlugins);
//               const joinAndQuitPlugin = joinAndQuitPlugins[0]
//               const {
//                 fundingGoal,
//                 minFeeToJoin,
//                 memberReputation
//               } = joinAndQuitPlugin.coreState.pluginParams;
//               try {
//                 await db.collection('daos').doc(id).set({
//                   id,
//                   address,
//                   memberCount,
//                   name,
//                   numberOfBoostedProposals,
//                   numberOfPreBoostedProposals,
//                   numberOfQueuedProposals,
//                   register,
//                   reputationId: reputation.id,
//                   reputationTotalSupply: parseInt(reputationTotalSupply),
//                   fundingGoal: fundingGoal.toString(),
//                   minFeeToJoin: minFeeToJoin.toString(),
//                   memberReputation: memberReputation.toString()
//                 })
//                 console.log(`[ Updated DAO ${name}@${address}] `);
//                 response.append(`updated DAO ${name}@${id}`)
//               } catch (error) {
//                 console.error(msg)
//                 const msg = `Failed to updated DAOs: ${error}`;
//                 throw Error(msg)
//               }
//             }
//           }
//           // if (dao.coreState.name.includes('Test DAO') || dao.coreState.name.includes('Car DAO')) {
//           //   console.log('DELETING: ', id)
//           //   await db.collection('daos').doc(id).delete();
//           // }
//         })
//       });
//   } catch(e) {
//     console.log('Error querying DAOs: ', e)
//     throw(e)
//   }
//   return '\n'.join(response)
// }

module.exports = {
  updateDaos,
  updatePlugins,
  updateProposals
}
