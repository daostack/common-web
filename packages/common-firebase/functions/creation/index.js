const functions = require('firebase-functions');
const express = require('express');
const cors = require('cors');
const bodyParser = require('body-parser');
const { responseExecutor, responseCreateExecutor } = require('../util/responseExecutor');
const { createCommon, createCommonTransaction } = require('./Common')
const { createRequestToJoinTransaction, createRequestToJoin  } = require('./RequestToJoin')
const { createVoteProposalTransaction, voteProposal  } = require('./VoteProposal')
const { createFundingProposalTransaction, createFundingProposal } = require('./FundingProposal');

const runtimeOptions = {
  timeoutSeconds: 540, // Maximum time 9 mins
}

const create = express();
create.use(bodyParser.json());
create.use(bodyParser.urlencoded({
  extended: true
}));
create.use(express.json());
create.use(express.urlencoded({ extended: true }));
create.use(cors({ origin: true }));

create.post('/createCommonTransaction', async (req, res) => {
  responseCreateExecutor(
    async () => {
      return await createCommonTransaction(req)
    },
    {
      req,
      res,
      successMessage: `Create common transaction successfully!`,
      errorMessage: `Unable to create a common transaction`
    }
  );
})

create.post('/createCommon', async (req, res) => {
  responseExecutor(
    async () => {
      return await createCommon(req);
    },
    {
      req,
      res,
      successMessage: `Common Created successfully!`,
      errorMessage: `Unable to create common!`
    }
  );
})

create.post('/createRequestToJoinTransaction', async (req, res) => {
  responseCreateExecutor(
    async () => {
      return await createRequestToJoinTransaction(req);
    },
    {
      req,
      res,
      successMessage: `Create requestToJoin transaction successfully!`,
      errorMessage: `Unable to create a request to join transaction!`
    }
  );
})

create.post('/createRequestToJoin', async (req, res) => {
  responseExecutor(
    async () => {
      return await createRequestToJoin(req);
    },
    {
      req,
      res,
      successMessage: `Request to join successfully!`,
      errorMessage: `Unable to create a request to join!`
    }
  );
})

  create.post('/createFundingProposalTransaction', async (req, res) => {
    responseCreateExecutor(
      async () => {
        return await createFundingProposalTransaction(req);
      },
      {
        req,
        res,
        successMessage: `Create funding proposal transaction successfully!!`,
        errorMessage: `Unable to create a funding proposal transaction`
      }
    );
})
  
  create.post('/createFundingProposal', async (req, res) => {
    responseExecutor(
      async () => {
        return await createFundingProposal(req);
      },
      {
        req,
        res,
        successMessage: `Create funding proposal successfully!`,
        errorMessage: `Unable to create a funding proposal !`
      }
    );
})

  create.post('/createVoteProposalTransaction', async (req, res) => {
    responseCreateExecutor(
      async () => {
        return await createVoteProposalTransaction(req);
      },
      {
        req,
        res,
        successMessage: `Create a vote proposal transaction successfully!`,
        errorMessage: `Unable to create vote transaction!`
      }
    );
  })
  
  create.post('/votePropoal', async (req, res) => {
    responseExecutor(
      async () => {
        return await voteProposal(req);
      },
      {
        req,
        res,
        successMessage: `Vote for the proposal successfully!`,
        errorMessage: `Unable to vote for a proposal !`
      }
    );
  })

exports.create = functions.runWith(runtimeOptions).https.onRequest(create);