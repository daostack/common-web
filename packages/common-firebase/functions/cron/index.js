const { backup } = require('./backupCron');

exports.crons = {
  backup
};