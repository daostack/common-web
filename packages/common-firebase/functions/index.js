require('module-alias/register');

const admin = require('firebase-admin');
const { databaseURL } = require('./settings');
const {env} = require('@env');

if(env.environment === 'dev') {
  admin.initializeApp();
} else {
  admin.initializeApp({
    credential: admin.credential.cert(require('@env/adminsdk-keys.json')),
    databaseURL: databaseURL
  });
}

const relayer = require('./relayer');
const graphql = require('./graphql');
const graphqlTriggers = require('./graphql/util/triggers');
const mangopay = require('./mangopay');
const mangopayTriggers = require('./mangopay/triggers');

// Add the __tests__ endpoints only if enabled
if(env.tests.enabled) {
  exports.tests = require('./tests').tests;
}

exports.relayer = relayer.relayer;
exports.graphql = graphql.graphql;
exports.mangopay = mangopay.mangopay;
exports.mangopaySubs = mangopayTriggers;
exports.graphqlSubs = graphqlTriggers;
// Disable notification
// exports.notification = require('./notification');

exports.cronJobs = require('./cron').crons;
