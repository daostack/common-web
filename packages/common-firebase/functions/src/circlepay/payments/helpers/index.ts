export { isFinalized, isSuccessful } from './statusHelper';

export { updatePayments } from './converter';

export { feesHelper } from './feesHelper';
export { failureHelper } from './failureHelper';