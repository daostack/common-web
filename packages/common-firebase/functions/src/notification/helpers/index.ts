export const getFundingRequestAcceptedTemplate = (country: string, amount: number) => {
  if (amount) {
    return !country
      ? 'userFundingRequestAcceptedUnknown'
      : (country === 'IL'
        ? 'userFundingRequestAcceptedIsraeli'
        : 'userFundingRequestAcceptedForeign');
  }
  return 'userFundingRequestAcceptedZeroAmount';
};