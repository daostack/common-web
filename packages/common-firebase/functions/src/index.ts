import * as cron from './cron';
import * as event from './event';
import * as notification from './notification';
import * as circlepayTriggers from './circlepay/triggers';

import { circlepayApp } from './circlepay';
import { commonsApp } from './common';
import { proposalCrons, proposalsApp } from './proposals';

// --- Express apps
export const commons = commonsApp;
export const circlepay = circlepayApp;
export const proposals = proposalsApp;

// --- Triggers and Subscriptions
exports.eventSub = event;
exports.notificationSub = notification;
exports.circlepayTriggers = circlepayTriggers;
exports.proposalCrons = proposalCrons;

exports.cronJobs = cron;
