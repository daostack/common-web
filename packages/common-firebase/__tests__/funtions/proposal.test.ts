import supertest from 'supertest';
import firebaseFunctionsTests from 'firebase-functions-test';

import * as functions from '../../functions/src';
import { getAuthToken } from '../helpers/auth';
import { proposalsApp } from '../helpers/supertests';
import { createTestCommon } from '../helpers/createTestCommon';
import exp = require('constants');

const proposal = supertest(functions.commons);

const joinEndpoint = '/create/join';
const fundingEndpoint = '/create/funding';

const test = firebaseFunctionsTests({
  projectId: 'common-tests'
});

const invalidJoinCreationPayload = {
  name: 'Hello, I\'m a test',
  description: 'I should not be created tho as I\'m incomplete'
};

const invalidFundingCreationPayload = {
  name: 'Hello, I\'m a test',
  description: 'I should not be created tho as I\'m incomplete'
};

const validJoinData = (commonId: string) => ({
  commonId,
  description: 'I wanna be a part',
  funding: 50000
});

const validFundingData = (commonId: string) => ({
  commonId,
  description: 'I wanna be a part',
  amount: 50000
});

describe('Proposal Related Cloud Functions', () => {
  afterAll(async () => {
    await test.cleanup();
  });

  it('should work', () => {
    expect(true).toBeTruthy();
  });

  it('should be healthy', async () => {
    const authToken = await getAuthToken('test-user');

    const response = await proposalsApp
      .get('/health')
      .set({
        Authorization: authToken
      });


    expect(response.status).toBe(200);
    expect(response.body.healthy).toBeTruthy();
  });

  describe('Join Proposal Creation', () => {
    it('should require authentication for join request creation', async () => {
      // Setup
      const common = await createTestCommon();

      // Act
      const response = await proposalsApp
        .post(joinEndpoint)
        .send(validJoinData(common.id));

      expect(response.status).toBe(401);
      expect(response.body.errorCode).toBe('AuthenticationError');
    });

    it('should fail validation on invalid input', async () => {
      // Setup
      const authToken = await getAuthToken('test-user');

      // Act
      const response = await proposalsApp
        .post(joinEndpoint)
        .send(invalidJoinCreationPayload)
        .set({
          Authorization: authToken
        });

      // Assert
      expect(response.status).toBe(422);
      expect(response.body.errorCode).toBeDefined();
      expect(response.body.errorCode).toBe('ValidationError');
      expect(response.body.data.errors).toMatchSnapshot();
    });

    it('should not allow founders to create join requests', async () => {
      // Setup
      const authToken = await getAuthToken('test-user');
      const common = await createTestCommon('test-user');

      // Act
      const response = await proposalsApp
        .post(joinEndpoint)
        .send(validJoinData(common.id))
        .set({
          Authorization: authToken
        });

      // Assert
      expect(response.status).toBe(400);
      expect(response.body.errorMessage).toBe('Cannot create join request for commons, that you are a member of');
    });

    it('should not allow to create join requests, if there is ongoing one', async () => {
      // Setup
      const authToken = await getAuthToken('test-user');
      const common = await createTestCommon('test-user');

      // Act
      const response = await proposalsApp
        .post(joinEndpoint)
        .send(validJoinData(common.id))
        .set({
          Authorization: authToken
        });

      // @todo
    });

    it('should make join request with valid input in commons, that you are not member of', async () => {
      // Setup
      const authToken = await getAuthToken('test-user');
      const common = await createTestCommon('another-test');

      // Act
      const response = await proposalsApp
        .post(joinEndpoint)
        .send(validJoinData(common.id))
        .set({
          Authorization: authToken
        });

      // Assert
      expect(response.body.message).toBe('Join request successfully created!');
      expect(response.body.proposerId).toBe('test-user');
      expect(response.body.type).toBe('join');
      expect(response.body.commonId).toBe(common.id);
    });
  });

  describe('Funding Proposal Creation', () => {
    it('should require authentication for funding proposals creation', async () => {
      // Setup
      const common = await createTestCommon();

      // Act
      const response = await proposalsApp
        .post(fundingEndpoint)
        .send(validJoinData(common.id));

      // Assert
      expect(response.status).toBe(401);
      expect(response.body.errorCode).toBe('AuthenticationError');
    });

    it('should not allow non members to create funding requests', async () => {
      // Setup
      const common = await createTestCommon('test-user');
      const authToken = await getAuthToken('test-user-not-member');

      // Act
      const response = await proposalsApp
        .post(fundingEndpoint)
        .send(validFundingData(common.id))
        .set({
          Authorization: authToken
        });

      // Assert
      expect(response.status).toBe(403);
      expect(response.body.error.includes('User tried to create funding request in common, that he is not part of')).toBeTruthy();
    });

    it('should fail on invalid input', async () => {
      // Setup
      const authToken = await getAuthToken('test-user-not-member');

      // Act
      const response = await proposalsApp
        .post(fundingEndpoint)
        .send(invalidFundingCreationPayload)
        .set({
          Authorization: authToken
        });

      // Assert
      expect(response.status).toBe(422);
      expect(response.body.data.errors).toMatchSnapshot();
    });

    it('should create funding request on valid data in commons, that you are a member of', async () => {
      // Setup
      const common = await createTestCommon('test-user');
      const authToken = await getAuthToken('test-user');

      // Act
      const response = await proposalsApp
        .post(fundingEndpoint)
        .send(validFundingData(common.id))
        .set({
          Authorization: authToken
        });

      // Assert
      expect(response.status).toBe(200);
      expect(response.body.message).toBe('Funding request successfully created!');

      expect(response.body.proposerId).toBe('test-user');
      expect(response.body.type).toBe('fundingRequest');
      expect(response.body.state).toBe('countdown');
    });
  });
});