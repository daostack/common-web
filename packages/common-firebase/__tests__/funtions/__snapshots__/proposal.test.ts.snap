// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Proposal Related Cloud Functions Funding Proposal Creation should fail on invalid input 1`] = `
Array [
  "commonId is a required field",
  "amount is a required field",
]
`;

exports[`Proposal Related Cloud Functions Join Proposal Creation should fail validation on invalid input 1`] = `
Array [
  "commonId is a required field",
  "funding is a required field",
]
`;
